check for WSAEMSGSIZE errors when receiving UDP packets on Windows

marten-seemann opened the pull request:
Fixes #3956. @thepaul, could you give this patch a try?

sukunrt reviewed (approved):
For windows this looks okay.
What happens on linux here? https://man7.org/linux/man-pages/man3/recvfrom.3p.html says
>If a message is too long to fit in the supplied
       buffer, and MSG_PEEK is not set in the flags argument, the excess
       bytes shall be discarded.
       
Is this fine? 

You might have to do handing for macOS since we merged #3946

thepaul commented:
> Fixes #3956. @thepaul, could you give this patch a try?

This does indeed fix the problem, according to my tests.

> You might have to do handling for macOS since we merged https://github.com/quic-go/quic-go/pull/3946

No changes should be necessary- the DF bit on outgoing datagrams shouldn't affect recvfrom(). It would affect things if we were using recvmsg() instead, and able to receive `MSG_TRUNC` control messages, but quic-go isn't there yet I think.

sukunrt commented:
@thepaul yes, there's no real change involved here. My comment was related to changes in `sys_conn_df.go` done in https://github.com/quic-go/quic-go/pull/3946/files 

That commit changed the build flags on `sys_conn_df.go` and introduced `sys_conn_df_darwin.go`. An older version of these files is modified in this PR.

thepaul commented:
Oh yes, I did read that change- sorry I wasn't clear! That change seems to amount to "set the DF bit on outgoing packets", and my claim is that that won't affect what is received with `recvfrom()`.

sukunrt commented:
This PR splits `isMsgSizeErr` into `isSendMsgSizeErr` and `isRecvMsgSizeErr`. That change needs to be made in this file too https://github.com/quic-go/quic-go/blob/master/sys_conn_df_darwin.go

marten-seemann commented:
> This PR splits `isMsgSizeErr` into `isSendMsgSizeErr` and `isRecvMsgSizeErr`. That change needs to be made in this file too https://github.com/quic-go/quic-go/blob/master/sys_conn_df_darwin.go

I see. This PR was based on an older commit, that's why CI succeeded on `push`, but failed on `pull_request`. I rebased and fixed.

sukunrt commented on sys_conn_df_darwin.go:
```diff
@@ -47,9 +47,8 @@ func setDF(rawConn syscall.RawConn) (bool, error) {
-func isMsgSizeErr(err error) bool {
-	return errors.Is(err, unix.EMSGSIZE)
-}
+func isSendMsgSizeErr(error) bool { return false }
```
should this be `unix.EMSGSIZE`?

marten-seemann replied:
If I understand EMSGSIZE correctly, it only applies on the receive side: That's why I split `isMsgSizeErr` into a send and a receive part.

On Linux, when a large packet is received, the kernel will just return you whatever fits into the buffer. This is fine, since QUIC packets are authenticated, and too short packets will be discarded.

thepaul replied:
Respectfully, I think you might have that backwards- `sendmsg()` will return `EMSGSIZE` on both Darwin and Linux when the message exceeds what it thinks the MTU is. `recvmsg()` can also return `EMSGSIZE`, but it only does so for `iov_msglen <= 0` or `iov_msglen > IOV_MAX`. `MSG_TRUNC` only applies on the receive side though.

marten-seemann replied:
https://man7.org/linux/man-pages/man7/udp.7.html says:
>  By default, Linux UDP does path MTU (Maximum Transmission Unit) discovery.  This means the kernel will keep track of the MTU to a specific target IP address and return EMSGSIZE when a UDP packet write exceeds it.  When this happens, the application should decrease the packet size.

sukunrt replied:
My understanding was `EMSGSIZE` is returned when we write a datagram larger than the MTU and the DF bit is set. So, `isSendMsgSizeErr` should return true when we receive `unix.EMSGSIZE` just like it does in sys_conn_df_linux.go

On the receive side, we never get any indication on macOS and Linux and the received datagram is just truncated.

marten-seemann replied:
@sukunrt Sorry, I misunderstood what you were saying. I added the check for `isSendMsgSizeErr` on darwin.
